server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java

```diff
@@ -104,16 +105,72 @@ protected void resolveRequest(ClusterState state, InternalRequest request) {
     protected void asyncShardOperation(GetRequest request, ShardId shardId, ActionListener<GetResponse> listener) throws IOException {
         IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
         IndexShard indexShard = indexService.getShard(shardId.id());
-        if (request.realtime()) { // we are not tied to a refresh cycle here anyway
-            super.asyncShardOperation(request, shardId, listener);
+        var shardRouting = clusterService.state().getRoutingNodes().node(clusterService.localNode().getId()).getByShardId(shardId);
+        if (shardRouting == null) {
+            listener.onFailure(new ShardNotFoundException(shardId, "shard is no longer assigned to current node"));
+            return;
+        }
+        // A TransportGetAction always goes to a search shard in Stateless.
+        if (shardRouting.isPromotableToPrimary() == false) {
+            handleGetOnUnpromotableShard(request, indexShard, listener);
         } else {
-            indexShard.awaitShardSearchActive(b -> {
-                try {
-                    super.asyncShardOperation(request, shardId, listener);
-                } catch (Exception ex) {
-                    listener.onFailure(ex);
-                }
-            });
+            assert DiscoveryNode.isStateless(clusterService.getSettings()) == false;
+            if (request.realtime()) { // we are not tied to a refresh cycle here anyway
+                super.asyncShardOperation(request, shardId, listener);
+            } else {
+                indexShard.awaitShardSearchActive(b -> {
+                    try {
+                        super.asyncShardOperation(request, shardId, listener);
+                    } catch (Exception ex) {
+                        listener.onFailure(ex);
+                    }
+                });
+            }
+        }
+    }
+
+    private void handleGetOnUnpromotableShard(GetRequest request, IndexShard indexShard, ActionListener<GetResponse> listener)
+        throws IOException {
+        ShardId shardId = indexShard.shardId();
+        var node = clusterService.state()
+            .nodes()
+            .get(clusterService.state().routingTable().shardRoutingTable(shardId).primaryShard().currentNodeId());
```

Comment:
`indexShard.getReplicationGroup().getRoutingTable().primaryShard().currentNodeId()`

Does that work?

Reply:
I think you can only do that if you are a primary. `handleGetOnUnpromotableShard` runs always on a non-primary.